docs/structured-streaming-state-data-source.md

```diff
@@ -0,0 +1,248 @@
+title: State data source (Experimental) Guide in Structured Streaming
+license: |
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+
+State data source Guide in Structured Streaming (Experimental)
+
+## Overview
+
+State data source provides functionality to manipulate the state from the checkpoint.
+
+As of Spark 4.0, state data source provides the read functionality with a batch query. Additional functionalities including write is on the future roadmap.
+
+NOTE: This data source is under experimental - source options and the behavior (output) might be subject to change.
+
+## Reading state key-values from the checkpoint
+
+State data source enables reading key-value pairs from the state in the checkpoint, via running a separate batch query.
+Users can leverage the functionality to cover two major use cases described below:
+
+* Construct a test checking both output and the state. It is non-trivial to deduce the key-value of the state from the output, and having visibility of the state would be a huge win on testing.
+* Investigate an incident against stateful streaming query. If users observe the incorrect output and want to track how it came up, having visibility of the state would be required.
+
+Users can read an instance of state store, which is matched to a single stateful operator in most cases. This means, users can expect that they can read the entire key-value pairs in the state for a single stateful operator. 
+
+Note that there could be an exception, e.g. stream-stream join, which leverages multiple state store instances internally. The data source abstracts the internal representation away from users and
+provide the users friendly approach to read the state. See the section for stream-stream join for more details.
+
+### Creating a State store for Batch Queries (all defaults)
+
+<div class="codetabs">
+
+<div data-lang="python" markdown="1">
+{% highlight python %}
+
+df = spark \
+.readStream \
+.format("statestore") \
+.load("<checkpointLocation>")
+
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+
+val df = spark
+.readStream
+.format("statestore")
+.load("<checkpointLocation>")
```

Comment:
Isn't this passed as a query option (as path) ?

Reply:
Can we add the option path example too ?

Reply:
Let's just explicitly mention in the explanation of source option. We'll probably have to replicate for all languages if we go with code example.